ogr/ogrsf_frmts/adbc/ogradbcdataset.cpp

```diff
@@ -0,0 +1,381 @@
+        error.clear();
+        AdbcStatementRelease(statement.get(), error);
+        return nullptr;
+    }
+
+    auto stream = std::make_unique<OGRArrowArrayStream>();
+    int64_t rows_affected = -1;
+    if (AdbcStatementExecuteQuery(statement.get(), stream->get(),
+                                  &rows_affected, error) != ADBC_STATUS_OK)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "AdbcStatementExecuteQuery() failed: %s", error.message());
+        error.clear();
+        AdbcStatementRelease(statement.get(), error);
+        return nullptr;
+    }
+
+    ArrowSchema schema = {};
+    if (stream->get_schema(&schema) != 0)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "get_schema() failed");
+        AdbcStatementRelease(statement.get(), error);
+        return nullptr;
+    }
+
+    return std::make_unique<OGRADBCLayer>(
+        this, pszLayerName, std::move(statement), std::move(stream), &schema);
+}
+
+/************************************************************************/
+/*                             ExecuteSQL()                             */
+/************************************************************************/
+
+OGRLayer *OGRADBCDataset::ExecuteSQL(const char *pszStatement,
+                                     OGRGeometry *poSpatialFilter,
+                                     const char *pszDialect)
+{
+    if (pszDialect && pszDialect[0] != 0 && !EQUAL(pszDialect, "NATIVE"))
+    {
+        return GDALDataset::ExecuteSQL(pszStatement, poSpatialFilter,
+                                       pszDialect);
+    }
+
+    auto poLayer = CreateLayer(pszStatement, "RESULTSET");
+    if (poLayer && poSpatialFilter)
+    {
+        if (poLayer->GetGeomType() == wkbNone)
+            return nullptr;
+        poLayer->SetSpatialFilter(poSpatialFilter);
+    }
+    return poLayer.release();
+}
+
+/************************************************************************/
+/*                                Open()                                */
+/************************************************************************/
+
+bool OGRADBCDataset::Open(const GDALOpenInfo *poOpenInfo)
+{
+    OGRADBCError error;
+
+    if (AdbcDatabaseNew(&m_database, error) != ADBC_STATUS_OK)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "AdbcDatabaseNew() failed: %s",
+                 error.message());
+        return false;
+    }
+
+    const char *pszFilename = poOpenInfo->pszFilename;
+    std::unique_ptr<GDALOpenInfo> poTmpOpenInfo;
+    if (STARTS_WITH(pszFilename, "ADBC:"))
+    {
+        pszFilename += strlen("ADBC:");
+        poTmpOpenInfo =
+            std::make_unique<GDALOpenInfo>(pszFilename, GA_ReadOnly);
+        poTmpOpenInfo->papszOpenOptions = poOpenInfo->papszOpenOptions;
+        poOpenInfo = poTmpOpenInfo.get();
+    }
+    const char *pszADBCDriverName =
+        CSLFetchNameValue(poOpenInfo->papszOpenOptions, "ADBC_DRIVER");
```

Comment:
> If at all possible, in R it would help to be able to initialize via the init function address:

Hum, would you suggest passing a function pointer as an hexadecimal address as a string like "0xDEADBEEF" ? I believe that's technically undefined behaviour in C/C++ for function pointers, also that will work in most cases. However I remember some time ago to have experimenting with the [CHERI architecture](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/) and pointers are kind of "fat pointers" and can't be "instanciated" from their address. And there's a security issue in allowing to pass this. I could do that, but protected with a configuration option to enable it explicitly.

Reply:
> Would a C or C++ API hook to inject `AdbcLoadDriver` or similar be secure?

definitely. The slightly annoying part is that setting the hook should be done in GDAL core (as the ADBC driver may be a plugin, that cannot be directly in it), but that's not much a big deal.

Reply:
I've added a new public header gdal_adbc.h per 8d86043696960632c662345c3dff7066367eb651 that can be used to call a ``void GDALSetAdbcDriverInitFunc(AdbcDriverInitFunc init_func)`` to define the entry point.
This is only compile tested on my side. Actual functional testing would be appreciated

Reply:
@paleolimbot Cool! I've merged your changes with minor adaptations